Allow arbitrary network range

ledyba-z opened the pull request:
to:
cc: @subspacecommunity/subspace-maintainers
related to:
resolves: #101, also resolves #63 

## Background

In current implementation, it is assumed that network masks are /24 for v4 and /112 for v6.
And also, the num of devices is limited up to 250.

In other words, you can set `SUBSPACE_IPV4_POOL` to `10.99.0.0/16`, but subspace just use only `10.99.0.1~10.99.0.254`.

In this PR, this limitation is removed.

### Changes

* Create `calc-default-gateway` command to calc the default gateway.
* Generate IP addresses using `net` standard library.
* `maxProfiles` constant is removed, because we will get an error if there is address to assign.

## Testing

I set up my own subspace server and create some devices.

ledyba-z commented:
Hmm... it looks there is no way to suppress `Make sure using this hardcoded IP address is safe here.` warnings for IPv6.
I tried link-local address and unique-local address, but I can't persuade sonar cloud.

ledyba-z commented:
Here is a dirty hack to suppress the warnings.

https://github.com/subspacecommunity/subspace/pull/108/commits/be04422775e934d7e024cb80c924abab757afac4

jack1902 commented on cmds/calc-default-gateway/main.go:
```diff
@@ -0,0 +1,43 @@
+}
+
+func main() {
+	var err error
+	flag.Parse()
+	if *help {
+		usage()
+		return
+	}
+	if flag.NArg() != 1 {
+		usage()
+		os.Exit(-1)
+	}
+	cidr := flag.Arg(0)
+	gw, err := util.CalcDefaultGateway(cidr)
+	if err != nil {
+		_, _ = fmt.Fprintf(os.Stderr, "Failed to calc default gw: %v\n", err)
+		os.Exit(-1)
+	}
+	fmt.Printf("%s\n", gw.String())
```
surely this can be implemented into the current main.go without having to be a completely seperate binary?

I state this as some are using the command `subspace` directly and this will not support them. I like the use of the `util` piece though, as i feel it can be used in the original `subspace` piece which will eventually have a golang standard project layout (its something i've adopted, as the format was layed out by others before me)

ledyba-z replied:
Yes, we still can use `subspace` command directly.

It it because this separated binary is for calculating default gateway in `entrypoint.sh` only used in Docker.

Here:
https://github.com/subspacecommunity/subspace/blob/46eb3bbaca1aa1325277b8edab63b5aaaee4a2db/entrypoint.sh#L49-L57

I googled some command to calculate the default gateway for arbitrary CIDR, but I can't find it.

ledyba-z replied:
OK, I will remove `util` package. Please wait a moment.

ledyba-z commented:
@jack1902 I removed `util` package. Could you take another look?

jack1902 commented:
@ledyba-z sorry for being unclear, what i meant was that the value is to calculated as a PART of subspace, not a seperate thing (i should be able to pass `192.168.0.1/24` into subspace and it will just work, without having to do some "external" calculation)

Thanks for the tests too :D ! Something this repo is in dire need of. Ideally the changes will only affect the go files and nothing should need to be changed in the `Makefile` or any of the `Docker` files (PR is open to move them which will drop this work)

jack1902 commented:
Also to add, in terms of the IP issue from Sonarcloud, i'll just ignore those as its not an issue at all (bit of an annoyance really)

ledyba-z commented:
I agree with you. I think it is not good behavior to calc default gateways implicitly(It's confusing).

Currently, there are ENV vars and command line options to configure subspace command.
I think it is also very confusing.
And also, there are 3 ENV vars to configure network address, netmask and gateway:

```
SUBSPACE_IPV4_GW
SUBSPACE_IPV4_POOL
SUBSPACE_IPV4_CIDR
```

Can I remove those variables?
(For docker environment, I will create `SUBSPACE_NETWORK_IPV4` variable and use it in entrypoint.sh to customize from docker-compose.yml.)

How about adding those flags to `subspace` command?
```
-nameserver 1.1.1.1
-listen-port=51820
-network-ipv4=192.168.0.1/24 ← 192.168.0.1 will be the gateway. GW address won't be calculated implicitly anymore
-network-ipv6=fc80::1/64
-enable-ipv6-nat=true
```

jack1902 commented:
Honestly, having the documentation just state that the first address in the given range will be the "gateway" is fine. Its normally the first or last useable address in a given network. I like the addition of being able to pass in:

```
--network-ipv4=CIDR_RANGE (192.168.0.0/24, 192.168.0.1 will be used as gateway)
```
same logic applies to ipv6

In regards to NAT being moved into subspace thats useful too, as it still allows the caller to disable it (Something i personally configure on a per client basis)

Are you on the Gophers slack? Should join the channel `#subspace` that way we can discuss this way quicker than PR comments on Github :).

These contributions will greatly improve this product too, especially if we port over to a golang web framework to bring structure :D

jack1902 commented:
I'm just creating a custom golang profile to remove the whole IP vuln thing as it's not an issue so removing the rule from sonarcloud

ledyba-z commented:
@jack1902 Thanks!

```
--network-ipv4=CIDR_RANGE (192.168.0.0/24, 192.168.0.1 will be used as gateway)
```

OK, I'll do so.

To remove those variable:
```
SUBSPACE_IPV4_GW
SUBSPACE_IPV4_POOL
SUBSPACE_IPV4_CIDR
```

I moved network configuration commands from entrypoint.sh to net_setup.go.

And also, I added `--enable-configure` and `--enable-dnsmasq` option for the users who don't use docker environment.

Could you take another look?

ledyba-z commented:
I also joined the slack channel, thanks to tell me that!

ledyba-z commented:
Sorry, I forgot to rewrite handler.go. Please wait a moment.

ledyba-z commented:
@jack1902 I fixed some remaining issues and also updated the README.md.
Please take a look!

(SonarCloud report tons of Code Smells, but I believe most of them can be ignored...)

jack1902 commented on go.mod:
```diff
@@ -8,6 +8,7 @@ require (
 	github.com/gorilla/securecookie v1.1.1
 	github.com/jteeuwen/go-bindata v3.0.8-0.20180305030458-6025e8de665b+incompatible
 	github.com/julienschmidt/httprouter v1.3.0
+	github.com/kevinburke/go-bindata v3.21.0+incompatible // indirect
```
drop this line, (its an issue with the Makefile)

jack1902 commented on go.sum:
```diff
@@ -18,6 +18,7 @@ github.com/jteeuwen/go-bindata v3.0.8-0.20180305030458-6025e8de665b+incompatible
 github.com/jteeuwen/go-bindata v3.0.8-0.20180305030458-6025e8de665b+incompatible/go.mod h1:JVvhzYOiGBnFSYRyV00iY8q7/0PThjIYav1p9h5dmKs=
 github.com/julienschmidt/httprouter v1.3.0 h1:U0609e9tgbseu3rBINet9P48AI/D3oJs4dN7jwJOQ1U=
 github.com/julienschmidt/httprouter v1.3.0/go.mod h1:JR6WtHb+2LUe8TCKY3cZOxFyyO8IZAc4RVcycCCAKdM=
+github.com/kevinburke/go-bindata v3.21.0+incompatible/go.mod h1:/pEEZ72flUW2p0yi30bslSp9YqD9pysLxunQDdb2CPM=
```
same as before, dropthis line

jack1902 commented on main.go:
```diff
@@ -92,6 +119,15 @@ func init() {
 	cli.BoolVar(&showHelp, "help", false, "display help and exit")
 	cli.BoolVar(&debug, "debug", false, "debug mode")
 	cli.StringVar(&semanticTheme, "theme", "green", "Semantic-ui theme to use")
+	cli.BoolVar(&enableNetworkSetup, "configure-network", false, "Configure networks automatically by subspace")
```
i'd expect this to be true by default

jack1902 commented on net.go:
```diff
@@ -0,0 +1,44 @@
+func generateIPAddr(v4Net *net.IPNet, v6Net *net.IPNet, id uint32) (net.IP, net.IP, error) {
+	v4 := cloneIP(v4Net.IP)
+	v6 := cloneIP(v6Net.IP)
+	for left, pos4, pos6 := id, len(v4)-1, len(v6)-2; left != 0; left, pos4, pos6 = left>>8, pos4-1, pos6-2 {
+		decimalId := byte(left & 0xff)
+		v4[pos4] += decimalId
+		hexId := uint16(decimalId%10) + uint16((decimalId/10)%10)*16 + uint16(decimalId/100)*256
+		v6[pos6+0] += byte((hexId >> 8) & 0xff)
+		v6[pos6+1] += byte(hexId & 0xff)
+	}
```
I have no idea what this means, seems like its very complex and i feel like it could be done easier? Thoughts @gavinelder

jack1902 commented on net_setup.go:
```diff
@@ -0,0 +1,118 @@
+  /sbin/iptables -t nat --append OUTPUT -s "{{.NetworkIPv4}}" -p tcp --dport 53 -j DNAT --to "{{.GatewayIPv4}}:53"
+fi
+
+# ipv6 - DNS Leak Protection
+if ! /sbin/ip6tables --wait -t nat --check OUTPUT -s "{{.NetworkIPv6}}" -p udp --dport 53 -j DNAT --to "{{.GatewayIPv6}}"; then
+  /sbin/ip6tables --wait -t nat --append OUTPUT -s "{{.NetworkIPv6}}" -p udp --dport 53 -j DNAT --to "{{.GatewayIPv6}}"
+fi
+if ! /sbin/ip6tables --wait -t nat --check OUTPUT -s "{{.NetworkIPv6}}" -p tcp --dport 53 -j DNAT --to "{{.GatewayIPv6}}"; then
+  /sbin/ip6tables --wait -t nat --append OUTPUT -s "{{.NetworkIPv6}}" -p tcp --dport 53 -j DNAT --to "{{.GatewayIPv6}}"
+fi
```
can we put this inside the IPV6NatEnabled piece, as not all boxes will have iptablesv6 installed

jack1902 commented on net_setup.go:
```diff
@@ -0,0 +1,118 @@
+sv restart dnsmasq
+{{end}}
+`, &v)
+}
```
add a newline to end of file

jack1902 commented on net_setup.go:
```diff
@@ -0,0 +1,118 @@
+if ip link show wg0 2>/dev/null; then
+  ip link del wg0
+fi
+ip link add wg0 type wireguard
```
shouldn't this be `ip link add dev wg0 type wireguard` ?

jack1902 commented on net_setup.go:
```diff
@@ -0,0 +1,118 @@
+  ip link del wg0
+fi
+ip link add wg0 type wireguard
+ip addr add "{{.GatewayIPv4WithCIDR}}" dev wg0
```
`ip address add dev wg0 GATEWAYWithCIDR`

These are the commands listed by wireguard themselves which is the reason i would prefer to use them

jack1902 commented on net_setup.go:
```diff
@@ -0,0 +1,118 @@
+fi
+ip link add wg0 type wireguard
+ip addr add "{{.GatewayIPv4WithCIDR}}" dev wg0
+ip addr add "{{.GatewayIPv6WithCIDR}}" dev wg0
```
See previous comment on the ip command preference. 

Also to add, have a look at https://www.wireguard.com/quickstart/

ledyba-z replied:
I copied those codes from previous `entrypoint.sh`... But I rewrote them according to your link. Thanks!

jack1902 commented on net_setup.go:
```diff
@@ -0,0 +1,118 @@
+
+# Create wireguard device
+if ip link show wg0 2>/dev/null; then
+  ip link del wg0
```
`ip link delete wg0`

jack1902 commented on net_setup.go:
```diff
@@ -0,0 +1,118 @@
+		GatewayIPv6WithCIDR: fmt.Sprintf("%s/%d", wireguardConfig.gatewayIPv6.String(), maskLenIPv6),
+		DnsmasqEnabled: enableDnsmasq,
+	}
+	return bash(`
```
i know this would extend the amount of work you have already put in, prehaps work for a future PR. Would it be possible to implement `go-iptables` by coreos for interactions with `ipTables` ?

ledyba-z replied:
OK, I will try using this library:
https://github.com/coreos/go-iptables

jack1902 commented on main.go:
```diff
@@ -92,6 +119,15 @@ func init() {
+	cli.StringVar(&nameserver, "nameserver", "1.1.1.1", "DNS server used by clients")
+	cli.StringVar(&networkIPv4, "network-ipv4", "10.99.97.0/24", "IPV4 network address to create. First one is reserved by server.")
+	cli.StringVar(&networkIPv6, "network-ipv6", "fd00::10:97:0/64", "IPV6 network address to create. First one is reserved by server.")
+	cli.StringVar(&endpointHost, "endpoint-host", "", "WireGuard  device's endpoint hostname. By default, http-host is used.")
```
you can set the default here for endpointHost to `httpHost` thus removing the need to check later on

ledyba-z replied:
I think if I change this line like:

```
cli.StringVar(&endpointHost, "endpoint-host", httpHost, "WireGuard  device's endpoint hostname. By default, http-host is used.")
```

httpHost is empty string in this line, so default value of `endpointHost` also will be empty string.

I moved the check code from `handler.go` to `main.go to` make it clear that default value of `endpointHost` is `httpHost`.

jack1902 commented on entrypoint.sh:
```diff
@@ -36,6 +38,14 @@ if [ -z "${SUBSPACE_LISTENPORT-}" ]; then
+fi
+
+if [ -z "${SUBSPACE_ALLOWED_IPS-}" ]; then
+  export SUBSPACE_ALLOWED_IPS="" # Use default values defined in subspace command
```
This needs to be set to `0.0.0.0/0, ::/0` if its not specified as the flag being passed in overrides the `go` default which means it comes out as blank

ledyba-z replied:
That's right! Thanks.

jack1902 commented on handlers.go:
```diff
@@ -358,73 +350,43 @@ func profileAddHandler(w *Web) {
 PrivateKey = ${wg_private_key}
 DNS = {{$.IPv4Gw}}, {{$.IPv6Gw}}
-Address = {{$.IPv4Pref}}{{$.Profile.Number}}/{{$.IPv4Cidr}},{{$.IPv6Pref}}{{$.Profile.Number}}/{{$.IPv6Cidr}}
+Address = {{$.IPv4Addr}}/{{$.IPv4Cidr}},{{$.IPv6Addr}}/{{$.IPv6Cidr}}
```
these should be:

`{{$.IPv4Addr}}/32,{{$.IPv6Addr}}/128`

jack1902 commented:
@ledyba-z i am aware that i have just opened #109 which will break parts of this PR (i wanted to fix this PR but i can't push to your fork)

jack1902 commented on cmd/subspace/net_setup.go:
```diff
@@ -0,0 +1,112 @@
+	if err != nil {
+		return "", err
+	}
+	setupNAT := func(network, gateway string) error {
+		err = iptable.AppendUnique("nat", "POSTROUTING", "-s", network, "-j", "MASQUERADE")
+		if err != nil {
+			return err
+		}
+		err = iptable.AppendUnique("filter", "FORWARD", "-m", "state", "--state", "RELATED,ESTABLISHED", "-j", "ACCEPT")
+		if err != nil {
+			return err
+		}
+		err = iptable.AppendUnique("filter", "FORWARD", "-s", network, "-j", "ACCEPT")
+		if err != nil {
+			return err
+		}
+		// DNS Leak Protection
+		err = iptable.AppendUnique("nat", "OUTPUT", "-s", network, "-p", "udp", "--dport", "53", "-j", "DNAT", "--to", fmt.Sprintf("%s:53", gateway))
+		if err != nil {
+			return err
+		}
+		err = iptable.AppendUnique("nat", "OUTPUT", "-s", network, "-p", "tcp", "--dport", "53", "-j", "DNAT", "--to", fmt.Sprintf("%s:53", gateway))
+		if err != nil {
+			return err
+		}
+		return nil
+	}
```
This can be a standalone function which will reduce the overall size of this one function

jack1902 replied:
Great work though on using `iptable` :D These contributions are very much appreciated!

ledyba-z replied:
OK, I made a new standalone function, `setupNAT`.

ledyba-z commented:
@jack1902 Thank you for your review!

I changed to use `nameserver` directly from the clients when `enable-dnsmasq=false`.
If someone wants to use another DNS server (such as onbound), he or she shall set `enable-dnsmasq=false` and `nameserver=(gateway v4), (gatewaty v6)`. Then subspace server and client should work as he or she expected (so I believe this PR can resolve #62 ).

Could you take another look?

(Sorry, I can't join the slack today...)

jack1902 commented on README.md:
```diff
@@ -96,19 +96,26 @@ $ subspace --http-host subspace.example.com
+| `endpoint-host`     | same as `http-host` | OPTIONAL: WireGuard device's endpoint hostname.                                                                           |
+| `listenPort`        | `51820`             | OPTIONAL: UDP port number for WireGuard device to listen                                                                  |
+| `enable-ipv6-nat`   | `true`              | OPTIONAL: If this and `configure-network` are enabled, `subspace` configure a NAT for IPv6 network.                       |
+| `allowed-ips`       | `0.0.0.0/0, ::/0`   | `IPv4/v6 CIDR list for client to connect via WireGuard VPN.`                                                              |
```
can we add `OPTIONAL: ` prefix to the description please?

ledyba-z replied:
OK!

jack1902 commented on README.md:
```diff
@@ -200,17 +212,20 @@ services:
-    - SUBSPACE_IPV4_GW=10.99.97.1
-    - SUBSPACE_IPV6_GW=fd00::10:97:1
-    - SUBSPACE_IPV6_NAT_ENABLED=1
+    - 'SUBSPACE_HTTP_HOST=subspace.example.org'
+    - 'SUBSPACE_BACKLINK=/'
+    - 'SUBSPACE_LETSENCRYPT=true'
+    - 'SUBSPACE_HTTP_INSECURE=false'
+    - 'SUBSPACE_HTTP_ADDR=:80'
+    - 'SUBSPACE_NAMESERVER=1.1.1.1'
+    - 'SUBSPACE_LISTENPORT=51820'
+    - 'SUBSPACE_ENDPOINT_HOST=you-can-use-another-hostname.subspace.example.com'
+    - 'SUBSPACE_NETWORK_IPV4=10.99.97.0/24'
+    - 'SUBSPACE_NETWORK_IPV6=fd00::10:97:0/64'
+    - 'SUBSPACE_IPV6_NAT_ENABLED=1'
+    - 'SUBSPACE_ENABLE_DNSMASQ=1'
+    - 'SUBSPACE_ALLOWED_IPS=0.0.0.0/0, ::/0'
+    - 'SUBSPACE_THEME=green'
```
can you remove the `'` as they aren't required in docker-compose and its confusing the diff

ledyba-z replied:
It is because if we wrote like:

```yaml
    - SUBSPACE_ALLOWED_IPS="0.0.0.0/0, ::/0"
```

I found that `allowedIPs` in golang will be `"0.0.0.0/0, ::/0"` (incldues quotes!), and it makes a syntax errors in generated config files.
I don't know this is a bug in docker-compose or expected behavior.

Could you tell me your idea about it?

jack1902 commented on cmd/subspace/main.go:
```diff
@@ -122,6 +158,11 @@ func main() {
 		os.Exit(1)
 	}
 
+	// Default value of endpointHost is the same as httpHost
+	if endpointHost == "" {
+		endpointHost = httpHost
+	}
+
```
can you update the cli commands so that the default value for endpointHost is `httpHost` as seen in a different comment. This way this check can be removed (`entrypoint.sh` will just have to set it as a default)

ledyba-z replied:
I did so, but I got this result.
Please see a screen shot:
<img width="868" alt="スクリーンショット 2020-07-06 15 55 18" src="https://user-images.githubusercontent.com/51456946/86570283-a3b33e00-bfaa-11ea-884c-31f8e440204e.png">

jack1902 replied:
prehaps have the default point to `&httpHost` (points to the memory address instead of the current value as `flags` hasn't parsed the values until its parsed them)

If that doesn't work then i'm happy to leave as is

ledyba-z replied:
Sorry to be late! (To tell you the truth, I was on a vacation!)

> prehaps have the default point to &httpHost (points to the memory address instead of the current value as flags hasn't parsed the values until its parsed them)

I think so too.

OK, I will leave it as is, thanks.

jack1902 commented on entrypoint.sh:
```diff
@@ -134,44 +88,35 @@ ListenPort = ${SUBSPACE_LISTENPORT}
+    # dnsmasq service log
+    mkdir -p /etc/service/dnsmasq/log/main
+    cat <<RUNIT >/etc/service/dnsmasq/log/run
+  #!/bin/sh
+  exec svlogd -tt ./main
```
please remove this change

ledyba-z replied:
This is just an indentation. Thanks, I will remove.

jack1902 commented on entrypoint.sh:
```diff
@@ -134,44 +88,35 @@ ListenPort = ${SUBSPACE_LISTENPORT}
-  cat <<RUNIT >/etc/service/dnsmasq/log/run
-#!/bin/sh
-exec svlogd -tt ./main
+    # dnsmasq service log
+    mkdir -p /etc/service/dnsmasq/log/main
+    cat <<RUNIT >/etc/service/dnsmasq/log/run
```
Reset this back to how it was (this diff isn't required)

ledyba-z replied:
OK.

jack1902 commented on entrypoint.sh:
```diff
@@ -134,44 +88,35 @@ ListenPort = ${SUBSPACE_LISTENPORT}
+  exec /usr/sbin/dnsmasq --no-daemon
 RUNIT
-  chmod +x /etc/service/dnsmasq/run
+    chmod +x /etc/service/dnsmasq/run
```
Reset this back to how it was (this diff isn't required)

jack1902 commented on entrypoint.sh:
```diff
@@ -134,44 +88,35 @@ ListenPort = ${SUBSPACE_LISTENPORT}
-  cat <<RUNIT >/etc/service/dnsmasq/run
-#!/bin/sh
-exec /usr/sbin/dnsmasq --no-daemon
+    mkdir -p /etc/service/dnsmasq
+    cat <<RUNIT >/etc/service/dnsmasq/run
+  #!/bin/sh
+  exec /usr/sbin/dnsmasq --no-daemon
```
Reset this back to how it was (this diff isn't required)

ledyba-z commented on cmd/subspace/net.go:
```diff
@@ -0,0 +1,44 @@
+		return nil, nil, fmt.Errorf("num of devices exceeds the limit of IP addr pool")
+	}
+	if !v6Net.Contains(v6) || !(v6.IsGlobalUnicast() || v6.IsLinkLocalUnicast()) {
+		return nil, nil, fmt.Errorf("num of devices exceeds the limit of IP addr pool")
```
OK, I  will optimize as possible.

ledyba-z commented on cmd/subspace/net_setup.go:
```diff
@@ -0,0 +1,113 @@
+	if err != nil {
+		return err
+	}
+	err = iptable.AppendUnique("nat", "POSTROUTING", "-s", network, "-j", "MASQUERADE")
```
Thanks, I will rewrote like that.
@jack1902 If possible, I will join @subspacecommunity and I will create branches in this repository. After that, you can push commits.

ledyba-z commented on entrypoint.sh:
```diff
@@ -36,6 +38,14 @@ if [ -z "${SUBSPACE_LISTENPORT-}" ]; then
 fi
 
+if [ -z "${SUBSPACE_ENDPOINT_HOST-}" ]; then
+  export SUBSPACE_ENDPOINT_HOST="" # Use default values defined in subspace command
```
It sounds good idea... But I found a corner case.

If we set `$SUBSPACE_HTTP_HOST` to `example.com:8080` (of course, this is a valid value!), `$SUBSPACE_ENDPOINT_HOST` contains port number, which should be invalid.

I propose that using host or IP address part of `$SUBSPACE_HTTP_HOST` for `--endpoint-host`.

To do so, I think we have to analyze `--http-host` param and set `endpointHost` in main.go.
I will send such a commit, so please consider about it.

ledyba-z commented:
It looks @jack1902 is currently busy now....

Could  you review this PR, if possible? > @gavinelder @agonbar @wizardels

This PR is so big, so it is difficult to rebase to resolve conflicts... 😅 

And also, if possible, I would like to join this org.

I'm using this software in the company I'm working for, so I think I would like to continuously contribute also in the future.

agonbar commented:
@ledyba-z I'm sorry, I wasn't so active lately, I finished the multiarch migration in dockerhub and finished the review of @jack1902 changes, I can move now to this PR. The changes are huge and will definitively need some time to give a look and accept the PR.

On the mean time one thing I can do is add you as developer to the project.

Once we merge this PR the focus of the project should be on improving CI with tests and migrate the code to a framework like fibers, cleaning code on the way. This discussion will be done mainly in the slack channel.

@ledyba-z don't hesitate to privately talk with any of us if we are behind on the reviews of PR, I think most of us use this project on our company but there's more things to do, so it's hard to invest all the time this project needs.

ledyba-z commented:
@agonbar Thank you so much to adding me to this community and reviewing! OK > discussing in the Slack channel.

> The changes are huge and will definitively need some time to give a look and accept the PR.

Of course, I will wait.

Currently, all of docker CI fail. However, I can't see the result... For example, I click the result of [386.dockerfile](https://hub.docker.com/login/?next=/redirect/%3Fresource_uri%3D/api/audit/v1/action/008cda24-4298-4a20-ad61-27f580115574/), but I got 404.

If I have to also join in some dockerhub's group to see the results, please invite [ledybaz](https://hub.docker.com/u/ledybaz).

After I'll be able to investigate why the CI fails, I'll work on them soon.

> @ledyba-z don't hesitate to privately talk with any of us if we are behind on the reviews of PR,

Thanks!

> I think most of us use this project on our company

Haha, I think so too. Subspace is very useful to use in companies (e.g. SAML)

> but there's more things to do, so it's hard to invest all the time this project needs.

Me too!

agonbar commented on README.md:
```diff
@@ -164,19 +171,24 @@ docker create \
+      # Optional variable to change WireGuard Listenport
+    --env 'SUBSPACE_LISTENPORT=51820' \
+    # Optinal variable to change the page to set the home button
+    --env 'SUBSPACE_BACKLINK=/'
```
There's a missing \ at the end of this line
```suggestion
    --env 'SUBSPACE_BACKLINK=/' \
```

ledyba-z commented:
Sorry to be late.

It is supported in latest master branch, so I will close.
Thank you all for reviewing!
